File: src/utils/frontmatter.ts
// Client-side frontmatter parse/stamp — keep the parsing logic in sync with
// server/src/blog.ts (parseFrontmatter / stampFrontmatter). Ported rather than
// shared because the SPA and hub are separate builds and it's ~80 lines.
//
// One extension over the server version: stampFrontmatter accepts string[]
// values (for tags) and serializes them as a YAML block list, replacing any
// existing scalar / inline-array / block-list form.

export interface Frontmatter {
  title?: string
  date?: string
  post?: boolean
  public?: boolean
  listed?: boolean
  log?: boolean
  project?: string
  status?: string
  tags?: string[]
}

// A plain (unquoted) YAML scalar can't hold everything a title can. The
// classic break: `title: Foo: bar` is a hard parse error ("incomplete explicit
// mapping pair"), which fails the Eleventy build for that post — so EVERY
// write path must go through yamlScalar(). Double-quoted style matches the
// convention already in the vault.
export function yamlScalar(value: string | boolean | number): string {
  if (typeof value !== 'string') return String(value)
  if (!needsYamlQuoting(value)) return value
  const escaped = value
    .replace(/\\/g, '\\\\')
    .replace(/"/g, '\\"')
    .replace(/\n/g, '\\n')
    .replace(/\r/g, '\\r')
    .replace(/\t/g, '\\t')
  return `"${escaped}"`
}

function needsYamlQuoting(s: string): boolean {
  if (s === '') return true
  if (s !== s.trim()) return true                        // leading/trailing space is dropped
  if (/^[-?:,[\]{}#&*!|>'"%@`]/.test(s)) return true     // YAML indicator char
  if (/:\s/.test(s) || s.endsWith(':')) return true      // reads as a nested mapping
  if (/\s#/.test(s)) return true                         // reads as a trailing comment
  if (/[\n\r\t]/.test(s)) return true
  // Would come back as a non-string
  if (/^(?:true|false|yes|no|on|off|null|~)$/i.test(s)) return true
  if (/^[+-]?(?:\d[\d_]*)?(?:\.\d*)?(?:[eE][+-]?\d+)?$/.test(s)) return true
  if (/^0[xob]/i.test(s)) return true
  return false
}

/**
 * Inverse of yamlScalar. Only strips a MATCHED quote pair — a bare trailing
 * quote (`Bruteforcing tailnet "fun names"`) is content, not a delimiter, and
 * the old `/^["']|["']$/g` strip silently ate it.
 */
export function unquoteYamlScalar(v: string): string {
  if (v.length >= 2 && v.startsWith('"') && v.endsWith('"')) {
    return v.slice(1, -1).replace(/\\(["\\/nrt])/g, (_, c: string) => (
      c === 'n' ? '\n' : c === 'r' ? '\r' : c === 't' ? '\t' : c
    ))
  }
  if (v.length >= 2 && v.startsWith("'") && v.endsWith("'")) {
    return v.slice(1, -1).replace(/''/g, "'")
  }
  return v
}

function isQuoted(v: string): boolean {
  return v.length >= 2
    && ((v.startsWith('"') && v.endsWith('"')) || (v.startsWith("'") && v.endsWith("'")))
}

export function parseFrontmatter(content: string): { fm: Frontmatter; body: string; raw: string } {
  const m = content.match(/^---\n([\s\S]*?)\n---\n?([\s\S]*)$/)
  if (!m) return { fm: {}, body: content, raw: '' }
  const raw = m[1]!
  const body = m[2] ?? ''
  const fm: Frontmatter = {}
  const lines = raw.split('\n')
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]!
    const kv = line.match(/^([a-zA-Z_][a-zA-Z0-9_-]*):\s*(.*)$/)
    if (!kv) continue
    const [, key, valRaw] = kv
    const val = valRaw!.trim()
    if (key === 'tags') {
      // Three forms in the wild:
      //   tags: foo            (scalar — single tag or comma/space-separated list)
      //   tags: [foo, bar]     (inline array)
      //   tags:\n  - foo\n     (block list)
      if (val.startsWith('[') && val.endsWith(']')) {
        fm.tags = val.slice(1, -1).split(',').map((s) => unquoteYamlScalar(s.trim())).filter(Boolean)
      } else if (val === '' || val === '[]') {
        const tags: string[] = []
        for (let j = i + 1; j < lines.length; j++) {
          const item = lines[j]!.match(/^\s*-\s+(.+?)\s*$/)
          if (!item) break
          tags.push(unquoteYamlScalar(item[1]!))
        }
        if (tags.length) fm.tags = tags
      } else if (isQuoted(val)) {
        // A quoted scalar is ONE tag, commas and all.
        const only = unquoteYamlScalar(val)
        if (only) fm.tags = [only]
      } else {
        const parts = (val.includes(',') ? val.split(',') : val.split(/\s+/))
          .map((s) => s.trim())
          .filter(Boolean)
        if (parts.length) fm.tags = parts
      }
      continue
    }
    if (val === '') continue
    // Bool coercion only for UNQUOTED values — `title: "true"` is the string.
    if (isQuoted(val)) (fm as Record<string, unknown>)[key!] = unquoteYamlScalar(val)
    else if (val === 'true') (fm as Record<string, unknown>)[key!] = true
    else if (val === 'false') (fm as Record<string, unknown>)[key!] = false
    else (fm as Record<string, unknown>)[key!] = val
  }
  return { fm, body, raw }
}

/**
 * Stamp keys into frontmatter, replacing existing lines or appending.
 * Array values (tags) serialize as a block list and replace ANY existing
 * form of the key (scalar / inline / block). Returns full file content.
 */
export function stampFrontmatter(
  content: string,
  updates: Record<string, string | boolean | string[]>,
): string {
  const m = content.match(/^---\n([\s\S]*?)\n---\n?([\s\S]*)$/)
  let raw = m ? m[1]! : ''
  const body = m ? (m[2] ?? '') : content
  let lines = raw.length ? raw.split('\n') : []

  for (const [k, v] of Object.entries(updates)) {
    // Remove any existing occurrence of the key. For block-list keys this
    // also removes the following `- item` lines.
    const idx = lines.findIndex((l) => l.match(new RegExp(`^${k}:`)))
    let insertAt = lines.length
    if (idx >= 0) {
      let end = idx + 1
      while (end < lines.length && lines[end]!.match(/^\s*-\s+/)) end++
      lines.splice(idx, end - idx)
      insertAt = idx
    }

    if (Array.isArray(v)) {
      const block = v.length ? [`${k}:`, ...v.map((item) => `  - ${yamlScalar(item)}`)] : [`${k}: `]
      lines.splice(insertAt, 0, ...block)
    } else {
      lines.splice(insertAt, 0, `${k}: ${yamlScalar(v)}`)
    }
  }
  raw = lines.join('\n')
  return `---\n${raw}\n---\n${body.startsWith('\n') ? body.slice(1) : body}`
}

/**
 * The byte range of the frontmatter block (including both `---` fences and
 * the trailing newline) in `content`, or null when there is none. Used to
 * surgically replace frontmatter via a CM6 transaction without touching the
 * body (preserves cursor + undo granularity).
 */
export function frontmatterRange(content: string): { from: number; to: number } | null {
  const m = content.match(/^---\n[\s\S]*?\n---\n?/)
  if (!m) return null
  return { from: 0, to: m[0]!.length }
}

/**
 * Public URL for a published post. Posts live in log/<name>.md or
 * projects/<slug>/log/<name>.md; both publish at /memo/log/<name>/.
 */
export function permalinkForLogPath(path: string): string | null {
  const m = path.match(/^projects\/[^/]+\/log\/([^/]+)\.md$/) ?? path.match(/^log\/([^/]+)\.md$/)
  if (!m) return null
  return `https://yousefamar.com/memo/log/${m[1]}/`
}

export const DRAFTS_DIR = 'log/drafts'
export const LEGACY_DRAFTS_DIR = 'scratch/blog-drafts'
export const LOG_DIR = 'log'

export function isDraftPath(path: string | null | undefined): boolean {
  return !!path && (
    path.startsWith(`${DRAFTS_DIR}/`)
    || path.startsWith(`${LEGACY_DRAFTS_DIR}/`)
    || /^projects\/[^/]+\/log\/drafts\/[^/]+\.md$/.test(path)
  )
}

/** Project slug implied by a draft's location, or null for unhomed drafts. */
export function projectForDraftPath(path: string | null | undefined): string | null {
  return path?.match(/^projects\/([^/]+)\/log\/drafts\/[^/]+\.md$/)?.[1] ?? null
}

export function isPublishedPath(path: string | null | undefined): boolean {
  return !!path && /^(?:projects\/[^/]+\/)?log\/[^/]+\.md$/.test(path)
}

File: android/app/src/main/kotlin/io/amar/console/data/notes/Frontmatter.kt
package io.amar.console.data.notes

/**
 * Client-side YAML-frontmatter parse/stamp — Kotlin port of the SPA's
 * src/utils/frontmatter.ts (itself kept in sync with server/src/blog.ts).
 * Ported rather than shared because the vault may be hub-local and the write
 * mode round-trips frontmatter through the editor buffer without a server hop.
 *
 * Only the small subset the writing-mode UI needs: title / tags / project +
 * the permalink mapping and draft/published detection.
 */
data class Frontmatter(
    val title: String? = null,
    val date: String? = null,
    val project: String? = null,
    val status: String? = null,
    val tags: List<String> = emptyList(),
    /** Everything else, verbatim scalar values (true/false/strings). */
    val extra: Map<String, String> = emptyMap(),
)

object FrontmatterParser {
    const val DRAFTS_DIR = "log/drafts"
    const val LEGACY_DRAFTS_DIR = "scratch/blog-drafts"
    const val LOG_DIR = "log"

    private val blockRe = Regex("^---\\n([\\s\\S]*?)\\n---\\n?([\\s\\S]*)$")
    private val kvRe = Regex("^([a-zA-Z_][a-zA-Z0-9_-]*):\\s*(.*)$")
    private val listItemRe = Regex("^\\s*-\\s+(.+?)\\s*$")

    private val indicatorRe = Regex("^[-?:,\\[\\]{}#&*!|>'\"%@`]")
    private val mappingRe = Regex(":\\s")
    private val commentRe = Regex("\\s#")
    private val boolNullRe = Regex("^(?:true|false|yes|no|on|off|null|~)$", RegexOption.IGNORE_CASE)
    private val numberRe = Regex("^[+-]?(?:\\d[\\d_]*)?(?:\\.\\d*)?(?:[eE][+-]?\\d+)?$")
    private val radixRe = Regex("^0[xob]", RegexOption.IGNORE_CASE)
    private val escapeRe = Regex("\\\\([\"\\\\/nrt])")

    /**
     * A plain (unquoted) YAML scalar can't hold everything a title can. The
     * classic break: `title: Foo: bar` is a hard parse error ("incomplete
     * explicit mapping pair"), which fails the Eleventy build for that post —
     * so EVERY write path goes through this. Double-quoted style matches the
     * convention already in the vault.
     */
    fun yamlScalar(value: Any?): String {
        if (value !is String) return value.toString()
        if (!needsQuoting(value)) return value
        val escaped = value
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t")
        return "\"$escaped\""
    }

    private fun needsQuoting(s: String): Boolean {
        if (s.isEmpty()) return true
        if (s != s.trim()) return true                              // leading/trailing space is dropped
        if (indicatorRe.containsMatchIn(s)) return true              // YAML indicator char
        if (mappingRe.containsMatchIn(s) || s.endsWith(":")) return true  // reads as a nested mapping
        if (commentRe.containsMatchIn(s)) return true                // reads as a trailing comment
        if (s.any { it == '\n' || it == '\r' || it == '\t' }) return true
        // Would come back as a non-string
        if (boolNullRe.matches(s)) return true
        if (numberRe.matches(s)) return true
        if (radixRe.containsMatchIn(s)) return true
        return false
    }

    /**
     * Inverse of [yamlScalar]. Only strips a MATCHED quote pair — a bare
     * trailing quote (`Bruteforcing tailnet "fun names"`) is content, not a
     * delimiter, and the old `trim('"', '\'')` silently ate it.
     */
    fun unquote(v: String): String {
        if (v.length >= 2 && v.startsWith("\"") && v.endsWith("\"")) {
            return escapeRe.replace(v.substring(1, v.length - 1)) { m ->
                when (m.groupValues[1]) {
                    "n" -> "\n"; "r" -> "\r"; "t" -> "\t"
                    else -> m.groupValues[1]
                }
            }
        }
        if (v.length >= 2 && v.startsWith("'") && v.endsWith("'")) {
            return v.substring(1, v.length - 1).replace("''", "'")
        }
        return v
    }

    private fun isQuoted(v: String): Boolean = v.length >= 2
        && ((v.startsWith("\"") && v.endsWith("\"")) || (v.startsWith("'") && v.endsWith("'")))

    /** Parse the frontmatter block; returns fm + body (never null). */
    fun parse(content: String): Frontmatter {
        val m = blockRe.find(content) ?: return Frontmatter()
        val raw = m.groupValues[1]
        val lines = raw.split('\n')
        var title: String? = null
        var date: String? = null
        var project: String? = null
        var status: String? = null
        var tags: List<String> = emptyList()
        val extra = LinkedHashMap<String, String>()

        var i = 0
        while (i < lines.size) {
            val line = lines[i]
            val kv = kvRe.find(line)
            if (kv == null) { i++; continue }
            val key = kv.groupValues[1]
            val valRaw = kv.groupValues[2].trim()
            if (key == "tags") {
                tags = parseTags(valRaw, lines, i)
                i++
                continue
            }
            if (valRaw.isEmpty()) { i++; continue }
            val v = unquote(valRaw)
            when (key) {
                "title" -> title = v
                "date" -> date = v
                "project" -> project = v
                "status" -> status = v
                else -> extra[key] = v
            }
            i++
        }
        return Frontmatter(title, date, project, status, tags, extra)
    }

    /** Just the tags list — used by the tag-autocomplete + meta bar. */
    fun parseTags(content: String): List<String> = parse(content).tags

    private fun parseTags(valRaw: String, lines: List<String>, keyIndex: Int): List<String> {
        // Three forms:  tags: foo   |   tags: [a, b]   |   tags:\n  - a\n  - b
        if (valRaw.startsWith("[") && valRaw.endsWith("]")) {
            return valRaw.substring(1, valRaw.length - 1).split(',')
                .map { unquote(it.trim()) }.filter { it.isNotEmpty() }
        }
        if (valRaw.isEmpty() || valRaw == "[]") {
            val out = ArrayList<String>()
            var j = keyIndex + 1
            while (j < lines.size) {
                val item = listItemRe.find(lines[j]) ?: break
                out.add(unquote(item.groupValues[1]))
                j++
            }
            return out
        }
        // A quoted scalar is ONE tag, commas and all.
        if (isQuoted(valRaw)) {
            val only = unquote(valRaw)
            return if (only.isEmpty()) emptyList() else listOf(only)
        }
        val parts = (if (valRaw.contains(',')) valRaw.split(',') else valRaw.split(Regex("\\s+")))
            .map { it.trim() }.filter { it.isNotEmpty() }
        return parts
    }

    /**
     * Stamp keys into frontmatter, replacing existing lines or appending; array
     * values (tags) serialize as a YAML block list, replacing ANY existing form
     * of the key. Returns full file content (with `---` fences).
     */
    fun stamp(content: String, updates: List<Pair<String, Any>>): String {
        val m = blockRe.find(content)
        var body = if (m != null) m.groupValues[2] else content
        val rawInit = if (m != null) m.groupValues[1] else ""
        val lines = if (rawInit.isNotEmpty()) ArrayList(rawInit.split('\n')) else ArrayList()

        for ((k, v) in updates) {
            val keyLineRe = Regex("^${Regex.escape(k)}:")
            val idx = lines.indexOfFirst { keyLineRe.containsMatchIn(it) }
            var insertAt = lines.size
            if (idx >= 0) {
                var end = idx + 1
                while (end < lines.size && listItemRe.containsMatchIn(lines[end])) end++
                repeat(end - idx) { lines.removeAt(idx) }
                insertAt = idx
            }
            when (v) {
                is List<*> -> {
                    val items = v.map { it.toString() }
                    val block = if (items.isNotEmpty()) listOf("$k:") + items.map { "  - ${yamlScalar(it)}" } else listOf("$k: ")
                    lines.addAll(insertAt, block)
                }
                else -> lines.add(insertAt, "$k: ${yamlScalar(v)}")
            }
        }
        val raw = lines.joinToString("\n")
        val cleanBody = if (body.startsWith("\n")) body.substring(1) else body
        return "---\n$raw\n---\n$cleanBody"
    }

    /** Byte range [from,to) of the frontmatter block (incl. fences + trailing NL), or null. */
    fun range(content: String): IntRange? {
        val re = Regex("^---\\n[\\s\\S]*?\\n---\\n?")
        val m = re.find(content) ?: return null
        if (m.range.first != 0) return null
        return 0 until m.value.length
    }

    /** log/<name>.md or projects/<slug>/log/<name>.md → https://yousefamar.com/memo/log/<name>/ ; else null. */
    fun permalinkForLogPath(path: String): String? {
        val m = Regex("^(?:projects/[^/]+/)?log/([^/]+)\\.md$").find(path) ?: return null
        return "https://yousefamar.com/memo/log/${m.groupValues[1]}/"
    }

    fun isDraftPath(path: String?): Boolean = path != null && (path.startsWith("$DRAFTS_DIR/") || path.startsWith("$LEGACY_DRAFTS_DIR/") || Regex("^projects/[^/]+/log/drafts/[^/]+\\.md$").matches(path))

    fun isPublishedPath(path: String?): Boolean =
        path != null && Regex("^(?:projects/[^/]+/)?log/[^/]+\\.md$").matches(path)

    /** Slug of a project index page (projects/<slug>.md or projects/<slug>/index.md). */
    fun projectSlugFromPath(path: String?): String? {
        if (path == null) return null
        val m = Regex("^projects/([^/]+?)(?:/index)?\\.md$").find(path) ?: return null
        return m.groupValues[1]
    }

    /** Slug of the project ENCLOSING a path (any file under projects/<slug>/…). */
    fun enclosingProjectSlug(path: String?): String? {
        if (path == null || !path.startsWith("projects/")) return null
        val rest = path.removePrefix("projects/")
        val slash = rest.indexOf('/')
        if (slash == -1) return rest.removeSuffix(".md").ifEmpty { null }
        return rest.substring(0, slash).ifEmpty { null }
    }

    /** Writing-mode chrome applies to drafts + published posts. */
    fun isWritingFile(path: String?): Boolean = isDraftPath(path) || isPublishedPath(path)
}
